src/stan/mcmc/chainset.hpp

```diff
@@ -0,0 +1,567 @@
+#ifndef STAN_MCMC_CHAINSET_HPP
+#define STAN_MCMC_CHAINSET_HPP
+
+#include <stan/io/stan_csv_reader.hpp>
+#include <stan/math/prim.hpp>
+#include <stan/analyze/mcmc/split_rank_normalized_ess.hpp>
+#include <stan/analyze/mcmc/split_rank_normalized_rhat.hpp>
+#include <boost/accumulators/accumulators.hpp>
+#include <boost/accumulators/statistics/stats.hpp>
+#include <boost/accumulators/statistics/mean.hpp>
+#include <boost/accumulators/statistics/tail_quantile.hpp>
+#include <boost/accumulators/statistics/p_square_quantile.hpp>
+#include <boost/accumulators/statistics/variance.hpp>
+#include <boost/accumulators/statistics/covariance.hpp>
+#include <boost/accumulators/statistics/variates/covariate.hpp>
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <sstream>
+#include <utility>
+#include <vector>
+#include <cstdlib>
+
+namespace stan {
+namespace mcmc {
+using Eigen::Dynamic;
+
+/**
+ * Checks that a Stan CSV file contains both a header row
+ * and a set of draws from the posterior.
+ * Throws exception if either are missing.
+ *
+ * @param stan_csv parsed csv file object
+ */
+void validate_sample(const stan::io::stan_csv& stan_csv) {
+  if (stan_csv.header.empty()) {
+    throw std::invalid_argument("Error: Stan CSV file missing header row");
+  }
+  if (stan_csv.samples.size() == 0) {
+    throw std::invalid_argument("Error: no sample found in Stan CSV file");
+  }
+}
+
+/**
+ * Reports the expected number of post-warmup draws in the CSV output file.
+ *
+ * @param stan_csv parsed csv file object
+ * @return expected number of draws
+ */
+size_t thinned_samples(const stan::io::stan_csv& stan_csv) {
+  size_t thinned_samples = stan_csv.metadata.num_samples;
+  if (stan_csv.metadata.thin > 1) {
+    thinned_samples = thinned_samples / stan_csv.metadata.thin;
+  }
+  return thinned_samples;
```

Comment:
Another place to consider if we want to enforce metadata being present -- I would prefer to default to the number of rows being the number of samples if metadata is absent